qfi_opt/spin_models.py

```diff
@@ -374,28 +372,40 @@ def act_on_subsystem(num_qubits: int, op: np.ndarray, *qubits: int) -> np.ndarra
 def get_jacobian_func(
     simulate_func: Callable,
     *,
-    disable_jax: bool = DISABLE_JAX,
-    step_sizes: float | Sequence[float] = 1e-4,
+    disable_diffrax: bool = DISABLE_DIFFRAX,
+    step_sizes: float | Sequence[float] = 1e-6,
 ) -> Callable:
     """Convert a simulation method into a function that returns its Jacobian."""
 
-    if not DISABLE_JAX:
-        jacobian_func = jax.jacrev(simulate_func, argnums=(0,), holomorphic=True)
-
-        def get_jacobian(*args: object, **kwargs: object) -> np.ndarray:
-            return jacobian_func(*args, **kwargs)[0]
+    if not DISABLE_DIFFRAX:
+
+        def get_jacobian(params: Sequence[float], *args: object, **kwargs: object) -> np.ndarray:
+            primals, vjp_func = jax.vjp(simulate_func, params, *args)
+            result = np.zeros((primals.shape[1], primals.shape[0], len(params)), dtype=COMPLEX_TYPE)
+            for i in range(primals.shape[1]):
+                for j in range(primals.shape[0]):
+                    seed = np.zeros(primals.shape, dtype=COMPLEX_TYPE)
+                    seed = seed.at[i, j].set(1.0)
```

Comment:
Should this be one of
```suggestion
            for i in range(primals.shape[0]):
                for j in range(primals.shape[1]):
                    seed = np.zeros(primals.shape, dtype=COMPLEX_TYPE)
                    seed = seed.at[i, j].set(1.0)
```
or

```suggestion
            for i in range(primals.shape[1]):
                for j in range(primals.shape[0]):
                    seed = np.zeros(primals.shape, dtype=COMPLEX_TYPE)
                    seed = seed.at[j, i].set(1.0)
```
?

For example, if `primals.shape` was (5, 3), I think the current would at some point try to set `seed.at[1, 4]`, which is an out-of-bounds index.

Reply:
@sriharikrishna I will assume that i and j should be swapped.  Please let me know if that's not the case.

Reply:
Yes, this is an error that went undetected, because the primals are square. I feel that the loop bounds should interchanged instead, not that it really matters.

```
for i in range(primals.shape[0]):
  for j in range(primals.shape[1]):
    seed = np.zeros(primals.shape, dtype=COMPLEX_TYPE)
    seed = seed.at[i, j].set(1.0)
```

Reply:
Yep, I went ahead with [this](https://github.com/jmlarson1/QFI-Opt/pull/39#discussion_r1486891648) (equivalent) resolution.